buildSrc/src/main/java/org/elasticsearch/gradle/internal/test/rerun/executer/RerunTestExecuter.java

```diff
@@ -0,0 +1,94 @@
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.gradle.internal.test.rerun.executer;
+
+import org.elasticsearch.gradle.internal.test.rerun.TestRerunTaskExtension;
+import org.gradle.api.GradleException;
+import org.gradle.api.internal.tasks.testing.JvmTestExecutionSpec;
+import org.gradle.api.internal.tasks.testing.TestDescriptorInternal;
+import org.gradle.api.internal.tasks.testing.TestExecuter;
+import org.gradle.api.internal.tasks.testing.TestResultProcessor;
+import org.gradle.internal.id.CompositeIdGenerator;
+import org.gradle.process.internal.ExecException;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public final class RerunTestExecuter implements TestExecuter<JvmTestExecutionSpec> {
+
+    private final TestRerunTaskExtension extension;
+    private final TestExecuter<JvmTestExecutionSpec> delegate;
+    private TestJvmCrashReporter jvmCrashReporter;
+
+    public RerunTestExecuter(TestRerunTaskExtension extension, TestExecuter<JvmTestExecutionSpec> delegate) {
+        this.extension = extension;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void execute(JvmTestExecutionSpec spec, TestResultProcessor testResultProcessor) {
+        int maxRetries = extension.getMaxRetries().get();
+        if (maxRetries <= 0) {
+            delegate.execute(spec, testResultProcessor);
+            return;
+        }
+
+        RerunTestResultProcessor retryTestResultProcessor = new RerunTestResultProcessor(testResultProcessor);
+
+        int retryCount = 0;
+        JvmTestExecutionSpec testExecutionSpec = spec;
+
+        while (true) {
+            try {
+                delegate.execute(testExecutionSpec, retryTestResultProcessor);
+                break;
+            } catch (ExecException e) {
+                if (retryCount == maxRetries) {
+                    throw new GradleException("Max retries hit", e);
+                } else {
+                    retryTestResultProcessor.reset(++retryCount == maxRetries);
+                }
+            }
+        }
+        jvmCrashReporter = new TestJvmCrashReporter(retryTestResultProcessor.getFailPaths());
+
+    }
+
+    @Override
+    public void stopNow() {
+        delegate.stopNow();
+    }
+
+    public void reportJvmCrashDetails() {
+        jvmCrashReporter.report();
+    }
+
+    private class TestJvmCrashReporter {
+        private List<TestDescriptorInternal> failPaths;
+
+        public TestJvmCrashReporter(List<TestDescriptorInternal> failPaths) {
+            this.failPaths = failPaths;
+        }
+
+        public void report() {
+            if(failPaths.size() > 0) {
+                String report = "================\n" +
+                        "Test JDK System exit trace:\n" +
```

Comment:
I think we should include in this message something to the effect of "test jvm exited unexpectedly" or would Gradle still report such a thing?

Reply:
It depends on the result of the tests.

If all tests pass in the first run and in the 2nd run (not taking the system exit causing one into account) then tests are only shown as one time executed in build scan (see example of the build scan I added to the infra test of this PR: https://gradle-enterprise.elastic.co/s/wsczh3rwdcvq4/tests)
If a test failed e.g. in the first execution and then succeed in the 2nd (again not taking the system exit causing one) then we see those tests marked as flaky (see https://gradle-enterprise.elastic.co/s/6vli7zag4ra7u/tests)

The test that's causing this issue is currently not yet reported at all as there was no fail test event received. 
I could potential tweak the rerun executor to synthesise a fail event for the system.exit causing test. Let me look into this.

Reply:
I tweaked the reporting a bit to include this. I think there's not much value to show the whole stack trace in gradle reporting the jvm exit as it doesn't provide any real value to trace this down.